File: Tests/UnitTestImpromptuInterface/PrivateTest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using ImpromptuInterface;
using ImpromptuInterface.Dynamic;
using ImpromptuInterface.InvokeExt;
using Microsoft.CSharp.RuntimeBinder;
using UnitTestSupportLibrary;


#if !SELFRUNNER
using NUnit.Framework;
#endif

#if SILVERLIGHT
namespace UnitTestImpromptuInterface.Silverlight
#else
namespace UnitTestImpromptuInterface
#endif
{
    [TestFixture]
    public class PrivateTest : Helper
    {
        [Test]
        public void TestExposePrivateMethod()
        {
            var tTest = new TestWithPrivateMethod();
            var tExposed = tTest.ActLike<IExposePrivateMethod>();
            Assert.AreEqual(3, tExposed.Test()); 
        }

        [Test]
        public void TestDoNotExposePrivateMethod()
        {
            var tTest = new TestWithPrivateMethod();
            var tNonExposed = tTest.WithContext(this).ActLike<IExposePrivateMethod>();
            AssertException<RuntimeBinderException>(() => tNonExposed.Test());
        }

        [Test]
        public void TestInvokePrivateMethod()
        {
            var tTest = new TestWithPrivateMethod();
            Assert.AreEqual(3, Impromptu.InvokeMember(tTest,"Test"));
        }

        [Test]
        public void TestInvokePrivateMethodAcrossAssemblyBoundries()
        {
            var tTest = new PublicType();
            Assert.AreEqual(true, Impromptu.InvokeMember(tTest, "PrivateMethod", 3));
        }

        [Test]
        public void TestInvokeInternalTypeMethodAcrossAssemblyBoundries()
        {
            var tTest = PublicType.InternalInstance;
            Assert.AreEqual(true, Impromptu.InvokeMember(tTest, "InternalMethod", 3));
        }

        [Test]
        public void TestInvokeDoNotExposePrivateMethod()
        {
            var tTest = new TestWithPrivateMethod();
            AssertException<RuntimeBinderException>(() => Impromptu.InvokeMember(tTest.WithContext(this), "Test"));
        }

        [Test]
        public void TestCacheableDoNotExposePrivateMethod()
        {
            var tTest = new TestWithPrivateMethod();
            var tCachedInvoke = new CacheableInvocation(InvocationKind.InvokeMember, "Test");
            AssertException<RuntimeBinderException>(() => tCachedInvoke.Invoke(tTest));
        }

        [Test]
        public void TestCacheableExposePrivateMethodViaInstance()
        {
            var tTest = new TestWithPrivateMethod();
            var tCachedInvoke = new CacheableInvocation(InvocationKind.InvokeMember, "Test", context: tTest);
            Assert.AreEqual(3, tCachedInvoke.Invoke(tTest));
        }

        [Test]
        public void TestCacheableExposePrivateMethodViaType()
        {
            var tTest = new TestWithPrivateMethod();
            var tCachedInvoke = new CacheableInvocation(InvocationKind.InvokeMember, "Test", context:typeof(TestWithPrivateMethod));
            Assert.AreEqual(3, tCachedInvoke.Invoke(tTest)); 
        }
    }

    public class TestWithPrivateMethod
    {
        private int Test()
        {
            return 3;
        }
    }


    public interface IExposePrivateMethod
    {
        int Test();
    }
}

File: Tests/UnitTestImpromptuInterface.VBNet/PrivateTest.vb
Option Strict Off

Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports NUnit.Framework
Imports ImpromptuInterface
Imports ImpromptuInterface.Dynamic
Imports ImpromptuInterface.InvokeExt
Imports Microsoft.CSharp.RuntimeBinder
Imports UnitTestSupportLibrary


Namespace VBNET

    <TestFixture()> _
    Public Class PrivateTest
        Inherits Helper
        <Test()> _
        Public Sub TestExposePrivateMethod()
            Dim tTest As TestWithPrivateMethod = New TestWithPrivateMethod()
            Dim tExposed = tTest.ActLike(Of IExposePrivateMethod)()
            Assert.AreEqual(3, tExposed.Test())
        End Sub

        <Test()> _
        Public Sub TestDoNotExposePrivateMethod()
            Dim tTest As TestWithPrivateMethod = New TestWithPrivateMethod()
            Dim tNonExposed = tTest.WithContext(Me).ActLike(Of IExposePrivateMethod)()
            AssertException(Of RuntimeBinderException)(Function() tNonExposed.Test())
        End Sub

        <Test()> _
        Public Sub TestInvokePrivateMethod()
            Dim tTest = New TestWithPrivateMethod()
            Assert.AreEqual(3, Impromptu.InvokeMember(tTest, "Test"))
        End Sub

        <Test()> _
        Public Sub TestInvokePrivateMethodAcrossAssemblyBoundries()
            Dim tTest = New PublicType()
            Assert.AreEqual(True, Impromptu.InvokeMember(tTest, "PrivateMethod", 3))
        End Sub

        <Test()> _
        Public Sub TestInvokeInternalTypeMethodAcrossAssemblyBoundries()
            Dim tTest = PublicType.InternalInstance
            Assert.AreEqual(True, Impromptu.InvokeMember(tTest, "InternalMethod", 3))
        End Sub

        <Test()> _
        Public Sub TestInvokeDoNotExposePrivateMethod()
            Dim tTest As TestWithPrivateMethod = New TestWithPrivateMethod()
            AssertException(Of RuntimeBinderException)(Function() Impromptu.InvokeMember(tTest.WithContext(Me), "Test"))
        End Sub

        <Test()> _
        Public Sub TestCacheableDoNotExposePrivateMethod()
            Dim tTest = New TestWithPrivateMethod()
            Dim tCachedInvoke = New CacheableInvocation(InvocationKind.InvokeMember, "Test")
            AssertException(Of RuntimeBinderException)(Function() tCachedInvoke.Invoke(tTest))
        End Sub

        <Test()> _
        Public Sub TestCacheableExposePrivateMethodViaInstance()
            Dim tTest = New TestWithPrivateMethod()
            Dim tCachedInvoke = New CacheableInvocation(InvocationKind.InvokeMember, "Test", context:=tTest)
            Assert.AreEqual(3, tCachedInvoke.Invoke(tTest))
        End Sub

        <Test()> _
        Public Sub TestCacheableExposePrivateMethodViaType()
            Dim tTest = New TestWithPrivateMethod()
            Dim tCachedInvoke = New CacheableInvocation(InvocationKind.InvokeMember, "Test", context:=GetType(TestWithPrivateMethod))
            Assert.AreEqual(3, tCachedInvoke.Invoke(tTest))
        End Sub
    End Class

    Public Class TestWithPrivateMethod
        Private Function Test() As Integer
            Return 3
        End Function
    End Class


    Public Interface IExposePrivateMethod
        Function Test() As Integer
    End Interface
End Namespace
